### PR TITLE
svcat : Allow dot syntax in --params

### LINKED ISSUE #1853: [svcat] Explode given parameters with dot syntax
Given, `svcat yada --param foo.bar=baz` it would be neat if that turned into an object like:

```
{
  foo: {
    bar: "baz"
  }
}
```

### PR DESCRIPTION
**Pull request by kikisdeliveryservice**
Closes #1853 

Add parsing of params to allow a param to hold value of map so that 
--param foo.bar.baz=boop -- > map{foo: map{bar: map{"baz":"boop"}

### CODE REVIEW

File: cmd/svcat/parameters/parameters.go [hunk 1]
```diff
@@ -56,19 +60,38 @@ func ParseVariableAssignments(params []string) (map[string]interface{}, error) {
...
+			}
 		} else {
 			switch storedValType := storedValue.(type) {
 			case string:
 				variables[variable] = []string{storedValType, value}
 			case []string:
 				variables[variable] = append(storedValType, value)
+			case map[string]string:
+				if len(subKey) > 0 {
+					varsv, submapv := reflect.ValueOf(variables[variable]), reflect.ValueOf(subMap)
```
Comment (staebler) [inline cmd/svcat/parameters/parameters.go]: I don't see what reflection is doing for us here? Is this different than doing the following?
```
for k, v := range sumMap {
   storedValType[k] = v
}
```
Ultimately, however, I don't think that we need to use a `subMap` at all since each loop through `params` is only going to have a single entry in `subMap`.

Comment (kikisdeliveryservice) [reply cmd/svcat/parameters/parameters.go]: hi @staebler thanks for your comments!! much appreciated.  

just to clarify that third case what is happening is:
storedValType[variable] = k[v] as opposed to storedValType[k]=v

so that storedValType[variable] can have values of ---> k[v], x[y], etc...

I went with the subMap, because the same k[v] map is used in two places (in if !ok and the 3rd case of the switch), so it seemed clearer to me how the k-v pair was being used that way. Will experiment and see what I can do with that this week. :)

Comment (staebler) [reply cmd/svcat/parameters/parameters.go]: `storedValType` already is `variables[variable]`, so you would not want to then use `storedValType[variable]`.
Also, `k[v]` does not make sense since `k` is the key of an entry in the map and `v` is the value of the same entry. In other words, `subMap[k]` is `v`. `k` is a string not a map.

Comment (kikisdeliveryservice) [reply cmd/svcat/parameters/parameters.go]: @staebler sorry, I meant to say that the final value of the third switch statement should be 
variables[variable] that holds the map subMap[subKey]= subValue so that we end up with : 
map {variable:{subKey:subValue}}

Comment (kikisdeliveryservice) [reply cmd/svcat/parameters/parameters.go]: @staebler Thanks for your comments, I'll work on this switch statement more this week!

**Comment by carolynvs** [comment]
@kikisdeliveryservice Let me know if you are blocked on something, or it's finished and we can remove the WIP flag and review it! 😀

**Comment by kikisdeliveryservice** [comment]
@carolynvs hi! i have some of it working, but i think we need to further define the specs of the acceptable params. was waiting until everyone was back from the f2f. made some initial questions in https://github.com/kubernetes-incubator/service-catalog/issues/1853 maybe we should chat over there?

**Comment by kikisdeliveryservice** [comment]
chatted with @carolynvs and what we came up with for this is the following:
- no max depth to the dot syntax
- `--param foo.bar=baz --param foo.fighters=everlong --param foo = pretender` is invalid syntax as it cannot both use dot syntax and hold a string as values.

**Comment by kikisdeliveryservice** [comment]
Taking off the WIP, since this is in a good state for a review & feedback.  =)

File: cmd/svcat/parameters/parameters.go [hunk 2]
```diff
@@ -56,19 +57,45 @@ func ParseVariableAssignments(params []string) (map[string]interface{}, error) {
...
+				}
+			}
 		} else {
 			switch storedValType := storedValue.(type) {
 			case string:
 				variables[variable] = []string{storedValType, value}
 			case []string:
 				variables[variable] = append(storedValType, value)
+			case map[string]interface{}:
+				return nil, fmt.Errorf("(%s) already exists as an exploding param", variable)
```
Comment (carolynvs) [inline cmd/svcat/parameters/parameters.go]: > exploding param

😂 That is awesome but I think "exploding param" may be confusing. What about this?

`was already used as an object path with the dot syntax and cannot be mixed with other formats`

Or something like that. What do you think?

File: cmd/svcat/parameters/parameters.go [hunk 3]
```diff
@@ -56,19 +57,45 @@ func ParseVariableAssignments(params []string) (map[string]interface{}, error) {
...
+		// if variable:value -> variable:[old value, newvalue]
+		// if variable:[value1, value2] -> variable:[value1, value2, newvalue]
+
 		if !ok {
-			variables[variable] = value // if there is no key, add key&value as string
+			if len(dotKeys) == 0 {
+				variables[variable] = value
+			} else {
+				for i := len(dotKeys[1:]); i > 0; i-- {
+					if i == len(dotKeys[1:]) {
```
Comment (carolynvs) [inline cmd/svcat/parameters/parameters.go]: OMG, it took me a couple minutes but I see what you are doing now, saving it from the bottom up in nested maps! 😊  👍

File: cmd/svcat/parameters/parameters_test.go [hunk 4]
```diff
@@ -77,6 +77,40 @@ func TestParseVariableAssignments_OneVariableThreeValues(t *testing.T) {
 		t.Fatalf("%s\nexpected:\n\t%v\ngot:\n\t%v\n", "one var three values", want, got)
 	}
 }
+func TestParseVariableAssignments_InvalidDotParams(t *testing.T) {
+	params := []string{"a.=e"}
```
Comment (carolynvs) [inline cmd/svcat/parameters/parameters_test.go]: This would benefit from using subtests, see below for an example. Then you can quickly add variations of this test for `.a=e` or `a.b..=e` just to cover all the edge cases.

https://github.com/kubernetes-incubator/service-catalog/blob/c51705f264350f54244b337ac8c87d71503c8731/cmd/svcat/svcat_test.go#L62-L76

File: cmd/svcat/parameters/parameters.go (hunk 2 above)
Comment (kikisdeliveryservice) [reply cmd/svcat/parameters/parameters.go]: That works!

File: cmd/svcat/parameters/parameters_test.go (hunk 4 above)
Comment (kikisdeliveryservice) [reply cmd/svcat/parameters/parameters_test.go]: Perfect, will do that - thank you!

File: cmd/svcat/parameters/parameters.go (hunk 3 above)
Comment (kikisdeliveryservice) [reply cmd/svcat/parameters/parameters.go]: Yes, the last element in the dot params always has the value "value", so I made that the initial  map and went through the rest of the params backwards to build the nested maps around it. Looking at it now, I think my if-else is unnecessary since it only will fall into the if on the first go-round. I will just do the assignment on the last element and then loop through the rest to build the nested map!

**Comment by kikisdeliveryservice** [comment]
Updated PR with @carolynvs 's changes, tidied up some of the comments and refactored some the nested mapping. =)

File: cmd/svcat/parameters/parameters.go [hunk 5]
```diff
@@ -69,33 +69,31 @@ func ParseVariableAssignments(params []string) (map[string]interface{}, error) {
...
+				// last element in dot params will have value "value", the rest of map nested
+				// using remaining elements
+				// Ex: A.B.C = Z --> map[A:map[B:map[C:Z]]]
+				variables[variable] = map[string]interface{}{dotKeys[len(dotKeys)-1]: value}
+				for i := len(dotKeys) - 2; i > 0; i-- {
+					variables[variable] = map[string]interface{}{dotKeys[i]: variables[variable]}
 				}
 			}
 		} else {
+			// if variable:value -> variable:[old value, newvalue]
```
Comment (carolynvs) [inline cmd/svcat/parameters/parameters.go]: nit: It would help if these comments were placed immediately before the case statement, instead of grouping them at the top

**Comment by kikisdeliveryservice** [comment]
To clarify after talking to @staebler : he would like the following functionality that would pass his suggested (not yet implemented) test in his post above.: 
“a.b=x” and “a.c=y” -->a: {b: x, c: y}

I'll work on this an ping when it's done. :)

( cc: @carolynvs @jeremyrickard )

File: cmd/svcat/parameters/parameters.go (hunk 5 above)
Comment (kikisdeliveryservice) [reply cmd/svcat/parameters/parameters.go]: I'll get this into my next update!

**Comment by MHBauer** [comment]
@kikisdeliveryservice friendly boop
/woof

**Comment by kikisdeliveryservice** [comment]
Thanks for the ping @MHBauer ! Hoping to have something to close this out soon. :)

**Comment by jberkhahn** [comment]
What's the status on this? we're going through the backlog to try and clean stuff up

**Comment by jberkhahn** [comment]
Due to lack of activity, we're closing this. If you still have an interest in this, please reopen with these changes.
